### PR TITLE
[REEF-1793] Add configuration options for local thread mode

### PR DESCRIPTION
**Pull request by interesaaat**
This addressed the issue by
  * Adding configuration options for local thread mode

JIRA:
  [REEF-1793](https://issues.apache.org/jira/browse/REEF-1793)

Pull request:
  This closes #

### CODE REVIEW

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java [hunk 1]
```diff
@@ -30,6 +30,9 @@
    * Indicates an Evaluator that runs on the CLR.
    */
   CLR,
+  /** Indicates an Evaluator that runs on a local thread.
+   */
+   LOCAL,
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java]: This seems like an orthogonal category of `EvaluatorType`. The others are about the runtime environment (JVM vs. CLR), while this one is about the REEF runtime it is on; We have a field for that in `EvaluatorInfo`, IIRC. Maybe this information should move there?

Comment (interesaaat) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java]: I am adding the LOCAL Evaluator type because I then use it in AllocatatedEvaluatorImpl#makeEvaluatorConfiguration to properly configure the thread mode. In the method the configuration is picked based on Evaluator type.

**Comment by markusweimer** [comment]
Can you please rebase to current `master`? We just merged a change that might just fix our appveyor build failures.

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java (hunk 1 above)
Comment (jwang98052) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java]: Currently we have Evalautor written in Java and .Net. EveluatorType indicates what runtime env. that the Evalautor is running on, Java Evaluator runs on JVM and C# Evaluator runs on CLR. Both can run either locally or on YARN. Not sure what LOCAL means here? Does the thread model only apply for local runtime?

Comment (interesaaat) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java]: Yes it only applies to local runtime and is a configuration option by which client, driver and evaluators are all run as thread inside the same process (JVM for the moment). This is more like  an option for the actual EvaluatorType and is currently used in `AllocatatedEvaluatorImpl#makeEvaluatorConfiguration` to generate the proper configuration.

**Comment by interesaaat** [comment]
I should have done the merge, but I see that 1 error still pops up.

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java (hunk 1 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java]: Can't `AllocatatedEvaluatorImpl` have `org.apache.reef.runtime.common.evaluator.parameters.LocalThreadMode` as a configuration parameter?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/parameters/LocalThreadMode.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+package org.apache.reef.runtime.common.evaluator.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * Flag used to set whether we are running in local thread mode or not.
+ */
+@NamedParameter(doc = "Flag for local thread mode.", default_value = LocalThreadMode.NO)
+public final class LocalThreadMode implements Name<String> {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/parameters/LocalThreadMode.java]: Why not use a `Boolean` for this?

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java (hunk 1 above)
Comment (interesaaat) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java]: Yes, this can work, but I have to add a parameter to `EvaluatorManager`.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/parameters/LocalThreadMode.java (hunk 2 above)
Comment (interesaaat) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/parameters/LocalThreadMode.java]: Yes. I found that other parameters (e.g., `ClientPresent`) used this design so I also embraced it. Do you prefer to go with Boolean?

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java (hunk 1 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/evaluator/EvaluatorType.java]: Sure, that is no problem, as it doesn't expand our `public` API.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/parameters/LocalThreadMode.java (hunk 2 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/parameters/LocalThreadMode.java]: Let's keep it with a string, as it really makes no difference to Tang.

**Comment by markusweimer** [comment]
LGTM, will merge.
